include/alloc-inl.h

```diff
@@ -673,78 +619,96 @@ static inline size_t next_pow2(size_t in) {
 
 }
 
-/* This function makes sure *size is > size_needed after call.
- It will realloc *buf otherwise.
- *size will grow exponentially as per:
- https://blog.mozilla.org/nnethercote/2014/11/04/please-grow-your-buffers-exponentially/
- Will return NULL and free *buf if size_needed is <1 or realloc failed.
- @return For convenience, this function returns *buf.
- */
-static inline void *maybe_grow(void **buf, size_t *size, size_t size_needed) {
+struct maybe_grow_buf {
 
-  /* No need to realloc */
-  if (likely(size_needed && *size >= size_needed)) { return *buf; }
+  size_t size;
+  u8     buf[0];
 
-  /* No initial size was set */
-  if (size_needed < INITIAL_GROWTH_SIZE) { size_needed = INITIAL_GROWTH_SIZE; }
+};
 
-  /* grow exponentially */
-  size_t next_size = next_pow2(size_needed);
+#define SIZE_OFFSET offsetof(struct maybe_grow_buf, buf)
 
-  /* handle overflow and zero size_needed */
-  if (!next_size) { next_size = size_needed; }
+/* Returs the container element to this ptr */
+static inline struct maybe_grow_buf *maybe_grow_buf(void *buf) {
 
-  /* alloc */
-  *buf = realloc(*buf, next_size);
-  *size = *buf ? next_size : 0;
+  return (struct maybe_grow_buf *)((u8 *)buf - SIZE_OFFSET);
 
-  return *buf;
+}
+
+/* Returs the current size of this allocated chunk (without the size_t needed to
+  store this info The size will be >= to the last size_needed passed to
+  maybe_grow */
+static inline size_t maybe_grow_bufsize(void *buf) {
+
+  if (!buf) { return 0; }
+  // The size is stored one size_t to the left of the ptr given to the user
+  return maybe_grow_buf(buf)->size;
 
 }
 
 /* This function makes sure *size is > size_needed after call.
  It will realloc *buf otherwise.
  *size will grow exponentially as per:
  https://blog.mozilla.org/nnethercote/2014/11/04/please-grow-your-buffers-exponentially/
- Will FATAL if size_needed is <1.
+ Will return NULL and free *buf if size_needed is <1 or realloc failed.
  @return For convenience, this function returns *buf.
  */
-static inline void *ck_maybe_grow(void **buf, size_t *size,
-                                  size_t size_needed) {
+static inline void *maybe_grow(void **buf, size_t size_needed) {
```

Comment:
that this is the function that actually increases a buffer is not clear in contrast to the other names.
maybe rename this to maybe_grow_need or _enlarge, or ...

Reply:
Thats what `grow` was supposed to say 😃
Probably we want a better naming scheme overall?

Reply:
could also be a different naming scheme. having grow in every function name might be misleading :)